app/controllers/payments_controller.rb

```diff
@@ -95,52 +162,61 @@ def webhook
+    payment.update_attribute(:status,
+                             Payment.order_to_payment_status(klarna_order['status']))
+    payment.successfully_completed
+
+    # Capture the order in Klarna (indicates the order has been filled and
+    # payment settlement can occur)
+    KlarnaService.capture_order(klarna_id, klarna_order['order_amount'])
+
+    klarna_order
   end
 
-  def log_hips_activity(activity, severity, payment_id, hips_id, exc=nil)
-    ActivityLogger.open(HIPS_LOG, 'HIPS_API', activity, false) do |log|
+  def log_klarna_activity(activity, severity, payment_id, klarna_id, exc=nil)
+    ActivityLogger.open(KLARNA_LOG, 'Payments', activity, false) do |log|
       log.record(severity, "Payment ID: #{payment_id}") if payment_id
-      log.record(severity, "HIPS ID: #{hips_id}") if hips_id
-      log.record(severity, "Exception class: #{exc.class}") if exc
-      log.record(severity, "Exception message: #{exc.message}") if exc
+      log.record(severity, "Klarna ID: #{klarna_id}") if klarna_id
+
+      [exc, exc&.cause].each do |this_exc|
+        next unless this_exc
+        log.record(severity, "Exception class: #{this_exc.class}") if exc
+        log.record(severity, "Exception message: #{this_exc.message}") if exc
+      end
     end
   end
 
-  def hips_order_urls(user_id, payment_id, company_id, payment_type)
+  def klarna_order_urls(payment_id, user_id, company_id, payment_type)
     urls = {}
-    urls[:success] = payment_success_url(user_id: user_id,
-                                         company_id: company_id,
-                                         id: payment_id,
-                                         payment_type: payment_type)
+    urls[:checkout] = payments_url(user_id: user_id, company_id: company_id,
+                                   type: payment_type)
 
-    urls[:error]   = payment_error_url(user_id: user_id,
-                                       company_id: company_id,
-                                       id: payment_id,
-                                       payment_type: payment_type)
+    urls[:confirmation] = payment_confirmation_url(id: payment_id,
+                                                   user_id: user_id,
+                                                   disable_language_change: true,
+                                                   klarna_id: '{checkout.order.id}')
 
-    urls[:webhook] = (SHF_WEBHOOK_HOST || root_url) + payment_webhook_path.sub('/en', '')
+    urls[:push] = (SHF_WEBHOOK_HOST || root_url) +
+                  payment_klarna_push_path(id: payment_id,
+                                           klarna_id: '{checkout.order.id}').sub('/en', '')
     urls
   end
+
+  def notify_slack_of_exception(exc, method_sym)
+    class_and_method = "#{self.class.name}\##{method_sym}"
+
+    [exc, exc.cause].each do |this_exc|
+      break unless this_exc
+      SHFNotifySlack.failure_notification(class_and_method, text: this_exc.message)
+    end
+  end
```

Comment:
I wonder if we want to add this method to `SHFNotifySlack`.

Reply:
This seems outside the scope/responsibility of `SHFNotifySlack`.  
 Its responsibility is to notify Slack, not to format or gather the information. 
`SHFNotifySlack` expects all of the information (exception, whatever) to be sent to it.  That has to take place before it gets sent to `SHFNotifySlack`.

I can see having a class that takes and exception, gets the info from it (perhaps as you've done) and logs and sends it to `SHFNotifySlack`.

Also:  (1) Why are you handling the exception and its cause like this?
(2) you should have an RSpec test -- that will explain it (as well as test it, obv.)
You are always doing that: logging and notifying slack; you can DRY up code by having 1 method handle those two calls.

Reply:
This is a convenient way to unpack the exception stack and post the whole "story" to slack.
The two methods have different signatures and a combined method would thus not make sense.

Reply:
We should have the same information (the cause/ the whole story) recorded to the log so that we never lose that information.

Does seem like a method that takes an exception and returns a string with the "whole story" would be a really good method to have and be available to the whole system.  (Then that 1 method could be called so it would be written out to where-ever is needed, i.e. the log, Slack,...)
Don't know if this would just be in ApplicationController, or where.
(Separate PR to be sure!)